src/utils.ts

```diff
@@ -107,10 +107,10 @@ export async function getBalance(
   return value;
 }
 
-export function uiToNative(amount: number, decimals: number): BN {
-  return new BN(Math.round(amount * Math.pow(10, decimals)));
+export function uiToNative(uiAmount: number, decimals: number): BN {
+  return new BN(Math.round(uiAmount * Math.pow(10, decimals)));
 }
 
-export function nativeToUi(amount: number, decimals: number): number {
-  return amount / Math.pow(10, decimals);
+export function nativeToUi(nativeAmount: BN, decimals: number): number {
+  return nativeAmount.toNumber() / Math.pow(10, decimals);
```

Comment:
Would this function fail if the `nativeAmount` is too big?

The BN.toNumber() function seems to be limited to MA_SAFE_INTEGER, aka `9,007,199,254,740,991`, which could be 9 million UXP in native unit.

```
BN.prototype.toNumber = function toNumber () {
    var ret = this.words[0];
    if (this.length === 2) {
      ret += this.words[1] * 0x4000000;
    } else if (this.length === 3 && this.words[2] === 0x01) {
      // NOTE: at this stage it is known that the top bit is set
      ret += 0x10000000000000 + (this.words[1] * 0x4000000);
    } else if (this.length > 2) {
      assert(false, 'Number can only safely store up to 53 bits');
    }
    return (this.negative !== 0) ? -ret : ret;
  };
```

Reply:
Yeah it would probably fail, it was the case before tho, however using this function allows us to create a proper implementation later 👌

Reply:
In my understanding, in the frontend app we are using `nativeI80F48ToUIExpensive`, I believe your credix/refacto PR made a switch to use `nativeToUi`. That may create an issue with big numbers (issue we don't have now because we use `nativeI80F48ToUIExpensive`).

Reply:
oh I see what you mean. Yeah that might be a problem. However note that most of the codebase in the frontend I think uses the non-expensive version. I think this was the reference before:

```tsx
import {
  I80F48 as UXDI80F48Expensive,
  nativeI80F48ToUi as uxdNativeI80F48ToUiExpensive,
  nativeToUi as uxdNativeToUi,
  uiToNative as uxdUiToNativeBNExpensive,
  uiToNative as uxdUiToNative,
} from "@uxd-protocol/uxd-client";

export function nativeI80F48ToUIExpensive(
  amount: UXDI80F48Expensive,
  decimals: number
): number {
  return uxdNativeI80F48ToUiExpensive(amount, decimals).toNumber();
}
```

Since mango removal nothing of the sort can work however. Let me write a more safer nativeToUi / uiToNative then

Reply:
Wow after in-depth testing looks like none of those functions was working properly in main branch either for decimals > 19 which will be a problem on EVM, fully re-implemented those to have infinite precision, this should be a lot better

Reply:
Until now in ethereum we have used the `ethers` packages and its utilities functions, like:

```
ethers.utils.formatEther(allowance)
```

Reply:
Yeah definitly true, but frontend uses nativeToUi and uiToNative when dealing with EVM amounts too in the UI.

To be honest this kind of makes sense since a precision-infinite integer handling is very useful cross-platform wise. But this means that currently you have >19 decimals when dealing with EVM values